Show help text on demand

domn1995 opened the pull request:
Notable changes:
   - Created a `CommandErrorDisplayOptions` flags enum for customizing error output between showing no output, showing the exception message, or showing the help text.
   - Created unit tests addressing the above.
   - Updated readme with example usage.

Note:
   - I did not update the version information. Wasn't too sure how you wanted to handle that. 

Closes #48

domn1995 commented on Readme.md:
````diff
@@ -430,6 +438,48 @@ Division by zero is not supported.
+}
+```
+
+If you don't want `CliFx` to display any errors upon throwing an exception, simply set the `CommandErrorDisplayOptions.NoError` flag in the exception like this:
````
```suggestion
If you don't want `CliFx` to display any errors upon throwing an exception, simply set the `CommandErrorDisplayOptions.None` flag in the exception like this:
```

domn1995 commented on CliFx/Exceptions/CommandException.cs:
```diff
@@ -11,6 +11,11 @@ public class CommandException : Exception
         private const int DefaultExitCode = -100;
 
+        /// <summary>
+        /// How to display 
```
```suggestion
        /// Gets the error display options for this exception.
```

Tyrrrz commented:
Thanks! I'll take a look at it when I have more time.

domn1995 commented on CliFx/CliApplication.cs:
```diff
@@ -143,6 +143,38 @@ await applicationSchema
+
+            if (commandErrorDisplayOptions.HasFlag(CommandErrorDisplayOptions.None))
+            {
+                return commandException.ExitCode;
```
```suggestion
				// Return early because we don't want to print anything if
				// the display option is set to 'None'.
                return commandException.ExitCode;
```

domn1995 commented on CliFx/CliApplication.cs:
```diff
@@ -162,21 +194,24 @@ await applicationSchema
-                    ? commandException.ExitCode
-                    : ex.HResult;
+                var exitCode = HandleCommandException(commandLineArguments, ce);
+                return ce.ExitCode;
```
```suggestion
                return exitCode;
```

Tyrrrz commented:
Hi, I finally found some time to look at it.
I see you have introduced an enum to configure whether the error/help is printed or not.

I think we can simplify it and just have the boolean flag for whether the help text should be printed.
Then, if the exception contains a message, show it as well. Otherwise, print stack trace only if help text is not shown.

This makes things simpler and seems to be reasonable in terms of user expectations.

domn1995 commented:
Yeah, I was weighing the pros and cons of enum vs bool and settled on enum for the following reasons: 
- There's a somewhat general consensus that enums are better than bools because they are more extensible and expressive. Think `string.Split("foo", StringSplitOptions.RemoveEmptyEntries)` vs `string.Split("foo", true)`.
- Allows users to optionally display an error message before the help text. This seems like a very useful feature to help clarify to the user why we're showing the help text and many command line tools work this way. Using git as an example:
```
> git clone
You must specify a repository to clone.

usage: git clone [<options>] [--] <repo> [<dir>]

    -v, --verbose         be more verbose
    -q, --quiet           be more quiet
    --progress            force progress reporting
    ...
```

Tyrrrz commented:
With the approach I suggested, we'd be able to show the error if it was provided. The error logic should remain the same, but if help text is also shown, then there's no reason to print the stack trace. Stack trace should be the last resort if there's no message and help text is not shown.

As for enum vs bool, you are right. However, I want to deliberately limit the number of permutations here to make it easier to reason about the framework and test it. I would prefer to have a good default behavior that covers 90% of use cases than 3 alternative workflows that cover 100% :)

domn1995 commented:
Yeah the extra complexity is definitely a minus. 

So with the `bool` solution, if the user doesn't want to display an error message before the help text, do we want to support both of these or just the first option? 
- `throw new CommandException(showHelp: true)`   
- `throw new CommandException(message: "", showHelp: true)`

Tyrrrz commented:
Yes, basically:

- `throw new CommandException(message: "Error message", showHelp: false)`

Prints just the error message, as it does now.

- `throw new CommandException(message: "Error message", showHelp: true)`

Prints the error message followed by help text.

- `throw new CommandException(message: null, showHelp: true)`

Prints just the help text, without error message

- `throw new CommandException(message: null, showHelp: false)`

Prints the stack trace, as it does now.

domn1995 commented:
Sounds good. I'll update the PR.

domn1995 commented:
Found an issue with the existing codebase as I was implementing this:

It looks like we'll never get a stack trace if we pass `null` into an `Exception`'s message. That's because the .NET `Exception.cs` code for the `Exception.Message` property checks if it is `null` and returns a default value if it is. See the code below:

https://github.com/microsoft/referencesource/blob/a7bd3242bd7732dec4aebb21fbc0f6de61c2545e/mscorlib/system/exception.cs#L136

I suggest we make `CommandException` and `CliFxException`'s constructor's `message` parameters non-nullable to prevent future bugs related to this behavior.

Additionally, the code in `CliApplication.cs` will never hit the `null` case because of the aforementioned. The consequence of all of this is that there's no clean way to implement your last two use cases from the reply above because `throw new CommandException(message: null, ...)` will always print `Exception of type 'CliFx.Exceptions.CommandException' was thrown.` with the current logic.

The only way to get a stack trace currently is to `throw new CommandException("", ...)`.

I don't see any specs testing the stack trace printing behavior you described, so I suggest creating an issue to implement those tests.

Now back to the original feature request, the `enum` solution is looking more attractive again, as it will allow us to implement showing just the help text, the error message and help text, or the stack trace based on an `ErrorDisplayOptions` enum and seems like the cleanest way to get original behavior that was intended.

How would you like to proceed? Also let me know if you want some more data.

Cheers!

Tyrrrz commented:
Ha, i did not know this.

There actually exists a spec for that, but it doesn't verify that what's printed is stack trace, just that something is printed in the event that no message was provided.

https://github.com/Tyrrrz/CliFx/blob/1dab27de551c8fcb5abd2656b5372c0091a77eb5/CliFx.Tests/ErrorReportingSpecs.cs#L60-L82

Regarding enum, I guess my main concern is that that API lets me throw an exception with an error message, but also configure it to not show that error message. In that case it's not clear what's the error message for. With the aforementioned approach it's more invariant: you provide a message -> it gets shown, otherwise it doesn't. Of course, .NET's behavior makes it a bit more annoying to do, but it's possible to extend the constructor so that it implicitly sets something like `IsErrorMessageProvided` which is evaluated as `!string.IsNullOrWhiteSpace(message)`.

Tyrrrz commented:
Cool, let me know when I should look at it again.

domn1995 commented:
Ready for you to take a look at!

**Notable changes:**
- Added a `BaseCliFxException` to remove code duplication between `CommandException` and `CliFxException` related to new help text logic. This also lets us optionally show help text when `CliFxException`s are thrown.
- Updated `CliFxApplication` with help text and error message handling logic.
- Created unit tests addressing the new help text features.
- Updated the existing stack trace unit test that wasn't behaving correctly.
- Updated the readme with details on customizing error reporting to show help text, or error message, and help text.
- Updated CliFx.Dummy project with examples of the aforementioned functionality.

Closes #49 and facilitates implementation of #16

Tyrrrz commented on CliFx.Tests/ErrorReportingSpecs.cs:
```diff
@@ -29,7 +29,10 @@ public async Task Command_may_throw_a_generic_exception_which_exits_and_prints_f
+                "System.Exception:",
+                "Kaput", "at", 
+                "CliFx.Tests");
             stdErrData.Length.Should().BeGreaterThan("Kaput".Length);
```
You can remove this line then.

domn1995 replied:
Will remove.

domn1995 replied:
So looking at this again, this was the test that could erroneously pass even though a stack trace wasn't printed. That's because the default exception message from .NET's `Exception` class returns a message of length greater than `"Kaput".Length`. My additional check that it contains those additional strings makes sure that we're actually looking at a stack trace output and not the default exception message in the case that behavior gets reintroduced in the future.

Tyrrrz commented on CliFx.Tests.Dummy/Commands/ShowErrorMessageThenHelpTextOnErrorCommand.cs:
```diff
@@ -0,0 +1,20 @@
+        "to remind you of my subcommands.")]
+    public class ShowErrorMessageThenHelpTextOnCommandExceptionCommand : ICommand
+    {
+        public async ValueTask ExecuteAsync(IConsole console) =>
```
Remove `async`.

domn1995 replied:
Will remove.

Tyrrrz commented on CliFx.Tests.Dummy/Commands/ShowHelpTextOnErrorCommand.cs:
```diff
@@ -0,0 +1,18 @@
+        "If you use just me, I print the help text to remind you of my subcommands.")]
+    public class ShowHelpTextOnCommandExceptionCommand : ICommand
+    {
+        public async ValueTask ExecuteAsync(IConsole console) =>
```
Remove `async`

domn1995 replied:
Will remove.

Tyrrrz commented on CliFx.Tests/HelpTextSpecs.cs:
```diff
@@ -241,6 +241,6 @@ public async Task Help_text_lists_environment_variable_names_for_options_that_ha
                 "-b|--option-b", "Environment variable:", "ENV_OPT_B"
             );
-        }
+        }        
```
Stray whitespace

domn1995 replied:
Will remove.

Tyrrrz commented on Readme.md:
```diff
@@ -35,18 +35,26 @@ An important property of CliFx, when compared to some other libraries, is that i
-- [Debug and preview mode](#debug-and-preview-mode)
-- [Reporting progress](#reporting-progress)
-- [Environment variables](#environment-variables)
+- [CliFx](#clifx)
+  - [Download](#download)
+  - [Features](#features)
+  - [Screenshots](#screenshots)
+  - [Usage](#usage)
+    - [Quick start](#quick-start)
+    - [Binding arguments](#binding-arguments)
+    - [Argument syntax](#argument-syntax)
+    - [Value conversion](#value-conversion)
+    - [Multiple commands](#multiple-commands)
+    - [Reporting errors](#reporting-errors)
+        - [Customizing Error Reporting](#customizing-error-reporting)
+    - [Graceful cancellation](#graceful-cancellation)
+    - [Dependency injection](#dependency-injection)
+    - [Testing](#testing)
+    - [Debug and preview mode](#debug-and-preview-mode)
+    - [Reporting progress](#reporting-progress)
+    - [Environment variables](#environment-variables)
+  - [Benchmarks](#benchmarks)
+  - [Etymology](#etymology)
```
Let's keep the table of contents only for usage

domn1995 replied:
Oops,  I think VSCode auto generated the whole table when I saved after adding my section. I'll fix.

Tyrrrz commented on Readme.md:
````diff
@@ -430,6 +438,34 @@ Division by zero is not supported.
 1337
 ```
 
+##### Customizing Error Reporting
````
I think this could just be added to the previous part ("Reporting errors") instead as an addition (i.e. "you can also choose to show help text, bla bla bla").

domn1995 replied:
Sounds good. I'll change.

Tyrrrz commented on .gitignore:
```diff
@@ -338,4 +338,5 @@ ASALocalRun/
-healthchecksdb
\ No newline at end of file
+healthchecksdb
+/CliFx.Tests.Dummy/Properties/launchSettings.json
```
Do we need this?

domn1995 replied:
Looks like `healthchecksdb` got added by VS automatically. Doesn't look like we use it, so I'll remove it so it doesn't clutter the file.
I'll also remove the `launchSettings.json` file and `.gitignore` entry.

Tyrrrz replied:
I was referring just to `/CliFx.Tests.Dummy/Properties/launchSettings.json`

domn1995 commented on CliFx/CliApplication.cs:
```diff
@@ -143,6 +143,50 @@ await applicationSchema
+        /// Handle <see cref="CommandException"/>s differently from the rest because we want to
+        /// display it different based on whether we are showing the help text or not.
+        /// </summary>
+        private int HandleCommandException(IReadOnlyList<string> commandLineArguments, CommandException commandException)
```
Missing return documentation.

Tyrrrz replied:
That's fine, the return value is obvious and for private methods it's not too important.

domn1995 commented on CliFx/CliApplication.cs:
```diff
@@ -162,21 +206,23 @@ await applicationSchema
-                    ? commandException.ExitCode
-                    : ex.HResult;
+                var exitCode = HandleCommandException(commandLineArguments, ce);
+                return ce.ExitCode;
```
Should return `exitCode` here.

domn1995 commented:
@Tyrrrz, all requested changes were made. Hopefully ready for a merge :)
